src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp

```diff
@@ -452,14 +444,10 @@ void ShenandoahOldGeneration::validate_transition(State new_state) {
       assert(heap->young_generation()->old_gen_task_queues() != nullptr, "Young generation needs old mark queues.");
       assert(heap->is_concurrent_old_mark_in_progress(), "Should be marking old now.");
       break;
-    case WAITING_FOR_EVAC:
-      assert(_state == IDLE || _state == MARKING, "Cannot have old collection candidates without first marking, state is '%s'", state_name(_state));
+    case EVACUATING:
+      assert(_state == WAITING_FOR_BOOTSTRAP || _state == MARKING, "Cannot have old collection candidates without first marking, state is '%s'", state_name(_state));
```

Comment:
FWIW, the ASCII art state-transition-diagram doesn't show the transition from WAITING_FOR_BOOTSTRAP directly into EVACUATING, without passing through an intermediate MARKING.

Reply:
Yes, that can only happen at the end of a _GLOBAL_ collection. This is explained in the comment above the diagram. I will add it to the diagram.

Reply:
Got it. One way to manage this and still get stronger checking during all, but in particular non-global, collections, might be to organize the state transition checks per collection type, and make assertions specific to the kind of collection in which the transition is happening/being checked. But perhaps that is overkill, and in any case not necessary in this PR.